### PR TITLE
Better defaults for PHP installations with old ICU lib

### LINKED ISSUE #2448: INTL_IDNA_VARIANT_UTS46 not defined having ICU < 4.6
**Guzzle version(s) affected**: 6.5.0

**Description**  
`IDN conversion failed` exception being thrown when dealing with domain name having regular ASCII characters (no IDN conversion required at all). Exception is thrown at https://github.com/guzzle/guzzle/blob/6.5.0/src/Client.php#L241

**How to reproduce**  
Requires old version of ICU on e.g. PHP 7.2 (these two things are independent - unfortunately).

**Possible Solution** 
Omit `INTL_IDNA_VARIANT_UTS46` in case it's not available - at https://github.com/guzzle/guzzle/blob/6.5.0/src/Client.php#L221

**Additional References**

See the following references and how it was handled in the TYPO3 CMS project
* https://github.com/TYPO3/TYPO3.CMS/commit/3a2b54a1b36b99b525cc04eccf20fd018a24e4a6#diff-cb4f296a39c2f11c2784e0e940745b82R193-R199
* https://forge.typo3.org/issues/87953 (additional description & system details)

### PR DESCRIPTION
**Pull request by alexeyshockov**
Fixes #2448

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
But what if the function was implemented by a pollyfill?

**Comment by alexeyshockov** [comment]
@GrahamCampbell, don't see any issues with a polyfill. If it implements `idn_to_ascii()`, then all should be fine. [`symfony/polyfill` implements both the function and `INTL_IDNA_VARIANT_UTS46` constant](https://github.com/symfony/polyfill/blob/master/src/Intl/Idn/bootstrap.php#L37), so `idn_conversion` option will be `true` by default.

**Comment by GrahamCampbell** [comment]
Maybe guzzle should always load the polyfill?

**Comment by alexeyshockov** [comment]
I think it will not help here. The original issue with `6.5.0` is about old version of ICU library (`intl` extension is available, just "partially"). So `symfony/polyfill` will not work in this case, because it discriminates things by just checking for `idn_to_ascii()` existence.

Creating a polyfill for exactly this case seems to be very challenging, and IMO Guzzle should not force any polyfills usage to the end developer.

File: src/Client.php [hunk 1]
```diff
@@ -267,7 +268,14 @@ private function configureDefaults(array $config)
...
 
         // idn_to_ascii() is a part of ext-intl and might be not available
-        $defaults['idn_conversion'] = function_exists('idn_to_ascii');
+        $defaults['idn_conversion'] = function_exists('idn_to_ascii')
+            // Old ICU versions don't have this constant, so we are basically stuck (see https://github.com/guzzle/guzzle/pull/2424
+            // and https://github.com/guzzle/guzzle/issues/2448 for details)
+            && (
+                defined('INTL_IDNA_VARIANT_UTS46')
+                ||
+                version_compare(PHP_VERSION, '7.2.0') < 0
```
Comment (brandonkelly) [inline src/Client.php]: This would be faster:

```suggestion
                PHP_VERSION_ID < 70200
```

**Comment by johncarlson21** [comment]
guys this needs to be merged like now.. after my update, now I have a non working script. I'm going to manually implement this fix.

**Comment by Mark-H** [comment]
This PR indeed fixes the issue I reported in #2458.

**Comment by sagikazarmark** [comment]
I assume this has to be forward ported to master as well, right?

**Comment by alexeyshockov** [comment]
@sagikazarmark, correct

**Comment by jonnott** [comment]
...so when is 6.5.1 shipping? ;)

**Comment by jonnott** [comment]
Can 6.5.1 be expedited ... ? Currently anyone running CentOS 6 has broken code if using Guzzle :(

**Comment by sagikazarmark** [comment]
@jonnott you can downgrade to 6.4 for the time being. 6.5.1 will be released today or tomorrow.

**Comment by sagikazarmark** [comment]
@alexeyshockov can you send this (or the appropriate version) patch to master as well?

**Comment by jonnott** [comment]
> @jonnott you can downgrade to 6.4 for the time being. 6.5.1 will be released today or tomorrow.

I temporarily added 'idn_conversion'=>false to all my Guzzle\Client instances instead!

**Pull request merged**